lxd/instance/drivers/driver_lxc.go

```diff
@@ -5360,6 +5373,187 @@ func (d *lxc) FileExists(path string) error {
 	return nil
 }
 
+// FileSocket returns a connection to the forkfile handler.
+func (d *lxc) FileSocket() (net.Conn, error) {
+	// Lock to avoid concurrent spawning.
+	spawnUnlock := locking.Lock(fmt.Sprintf("forkfile_%d", d.id))
+	defer spawnUnlock()
+
+	// Attempt to connect on existing socket.
+	forkfilePath := filepath.Join(d.LogPath(), "forkfile.sock")
+
+	forkfileAddr, err := net.ResolveUnixAddr("unix", forkfilePath)
+	if err != nil {
+		return nil, err
+	}
+
+	forkfileConn, err := net.DialUnix("unix", nil, forkfileAddr)
+	if err == nil {
+		// Found an existing server.
+		return forkfileConn, nil
+	}
+
+	// Check for ongoing operations (that may involve shifting).
+	operationlock.Get(d.Project(), d.Name()).Wait()
+
+	// Create the listener.
+	os.Remove(forkfilePath)
+	forkfileListener, err := net.ListenUnix("unix", forkfileAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	// Spawn forkfile in a Go routine.
+	chReady := make(chan error)
+	go func() {
+		// Lock to avoid concurrent running forkfile.
+		runUnlock := locking.Lock(fmt.Sprintf("forkfile-running_%d", d.id))
+		defer runUnlock()
+
+		// Setup reverter.
+		reverter := revert.New()
+		defer reverter.Fail()
+
+		// Always cleanup on failure.
+		reverter.Add(func() {
```

Comment:
They were all defers and I had to change to the current setup to avoid issues with the ordering, I'll take another look to see if I've since changed something that makes this irrelevant now.

Reply:
So I don't think that's correct.

In this case, the reverter is for:
```
forkfileListener.Close()
os.Remove(forkfilePath)
```

Which we want to run on failure but on success, we want them to run BEFORE other defers, which is why it's structured that way. If we made them all defers, then those would run very late, specifically after the `d.unmount()` defer in the stopped container case.

And that would effectively cause a situation where:
 - Existing forkfile exits
 - Sockets are closed + deleted
 - New client causes a new forkfile to be spawned (held by -running lock)
 - Unmount is run, taking up to 10-30s in some cases
 - Old forkfile function exits, calls defers which delete the socket from the new forkfile
 - New forkfile process moves forward (-running lock is gone), forkfile is spawned but its socket doesn't exist
 - Client attempts to connect and fails (No such file or directory)

Reply:
Adding a comment on the reverter.Add call (there already was a comment explaining the reason later in the function).

Reply:
OK thanks.